File: src/models/Particle.ts
export type ParticleType = 'proton' | 'neutron' | 'electron';
export type OrbitalType = 'none' | 's' | 'p' | 'd' | 'f' | 'g';

export interface ParticleProps {
    x: number;
    y: number;
    type: ParticleType;
    orbitalType?: OrbitalType;
    radius?: number;
    charge?: number;
    mass?: number;
    dx?: number;
    dy?: number;
}

export class Particle {
    x: number;
    y: number;
    type: ParticleType;
    orbitalType: OrbitalType;
    radius: number;
    charge: number;
    mass: number;
    dx: number;
    dy: number;

    constructor({
        x,
        y,
        type,
        orbitalType = 'none',
        radius,
        charge,
        mass,
        dx = (Math.random() - 0.5) * 2,
        dy = (Math.random() - 0.5) * 2
    }: ParticleProps) {
        this.x = x;
        this.y = y;
        this.type = type;
        this.orbitalType = orbitalType;
        this.radius = radius || (type === 'electron' ? 10 : 20);
        this.charge = charge || (type === 'proton' ? 1 : type === 'electron' ? -1 : 0);
        this.mass = mass || (type === 'electron' ? 0.0005 : 3.0);
        this.dx = dx;
        this.dy = dy;
    }

    update(canvasWidth: number, canvasHeight: number, damping: number): void {
        // Colisiones con bordes
        if (this.x + this.radius > canvasWidth) {
            this.x = canvasWidth - this.radius;
            this.dx *= -0.9;
        }
        if (this.x - this.radius < 0) {
            this.x = this.radius;
            this.dx *= -0.9;
        }
        if (this.y + this.radius > canvasHeight) {
            this.y = canvasHeight - this.radius;
            this.dy *= -0.9;
        }
        if (this.y - this.radius < 0) {
            this.y = this.radius;
            this.dy *= -0.9;
        }

        // Actualizar posición
        this.x += this.dx;
        this.y += this.dy;

        // Aplicar amortiguación
        this.dx *= damping;
        this.dy *= damping;
    }
} 

File: src/particle.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub enum ParticleType {
    Proton,
    Neutron,
    Electron,
}

#[derive(Debug, Clone)]
pub struct Particle {
    pub x: f64,
    pub y: f64,
    pub dx: f64,
    pub dy: f64,
    pub ptype: ParticleType,
    pub radius: f64,
    pub charge: f64,
    pub mass: f64,
    pub orbiting_around: Option<usize>,                           // Índice del protón alrededor del que orbita
    pub orbital_position: Option<(char, usize, usize, usize)>, // (subnivel, nivel, orbital_idx, posicion_en_orbital)
}

impl Particle {
    pub fn new(x: f64, y: f64, ptype: ParticleType, mass: f64) -> Self {
        let (radius, charge) = match ptype {
            ParticleType::Proton => (20.0, 1.0),
            ParticleType::Neutron => (20.0, 0.0),
            ParticleType::Electron => (10.0, -1.0),
        };

        Particle {
            x,
            y,
            dx: 0.0,
            dy: 0.0,
            ptype,
            radius,
            charge,
            mass,
            orbiting_around: None,
            orbital_position: None,
        }
    }

    pub fn update(&mut self, canvas_width: f64, canvas_height: f64, damping: f64) {
        // Actualizar posición
        self.x += self.dx;
        self.y += self.dy;

        // Colisiones elásticas con bordes (sin pérdida de energía)
        let reflection = match self.ptype {
            ParticleType::Electron => -1.0, // Colisión completamente elástica
            _ => -1.0, // Colisión completamente elástica
        };

        if self.x + self.radius > canvas_width {
            self.x = canvas_width - self.radius;
            self.dx *= reflection;
        }
        if self.x - self.radius < 0.0 {
            self.x = self.radius;
            self.dx *= reflection;
        }
        if self.y + self.radius > canvas_height {
            self.y = canvas_height - self.radius;
            self.dy *= reflection;
        }
        if self.y - self.radius < 0.0 {
            self.y = self.radius;
            self.dy *= reflection;
        }

        // Aplicar amortiguación mínima
        self.dx *= damping;
        self.dy *= damping;
    }

    pub fn distance_to(&self, other: &Particle) -> f64 {
        let dx = self.x - other.x;
        let dy = self.y - other.y;
        (dx * dx + dy * dy).sqrt()
    }
}
